traffic_ops/traffic_ops_golang/deliveryservice/safe.go

```diff
@@ -0,0 +1,128 @@
+	ds, ok := UpdateSafe(w, r)
+	if !ok {
+		return
+	}
+	api.WriteRespAlertObj(w, r, tc.SuccessLevel, "Deliveryservice safe update was successful.", []tc.DeliveryServiceNullableV13{ds.DeliveryServiceNullableV13})
+}
+
+func UpdateSafeV12(w http.ResponseWriter, r *http.Request) {
+	ds, ok := UpdateSafe(w, r)
+	if !ok {
+		return
+	}
+	api.WriteRespAlertObj(w, r, tc.SuccessLevel, "Deliveryservice safe update was successful.", []tc.DeliveryServiceNullableV12{ds.DeliveryServiceNullableV12})
+}
+
+// UpdateSafe updates the delivery service, writing any errors. Returns true on success, or false on error. If an error occured, it will be written to the client and logged appropriately, and the caller shouldn't write anything further. On success, the caller should write the delivery service response to the client.
+func UpdateSafe(w http.ResponseWriter, r *http.Request) (tc.DeliveryServiceNullable, bool) {
+	inf, userErr, sysErr, errCode := api.NewInfo(r, []string{"id"}, []string{"id"})
+	if userErr != nil || sysErr != nil {
+		api.HandleErr(w, r, inf.Tx.Tx, errCode, userErr, sysErr)
+		return tc.DeliveryServiceNullable{}, false
+	}
+	defer inf.Close()
+
+	dsID := inf.IntParams["id"]
+
+	userErr, sysErr, errCode = tenant.CheckID(inf.Tx.Tx, inf.User, dsID)
+	if userErr != nil || sysErr != nil {
+		api.HandleErr(w, r, inf.Tx.Tx, errCode, userErr, sysErr)
+		return tc.DeliveryServiceNullable{}, false
+	}
+
+	ds := tc.DeliveryServiceSafeUpdate{}
+	if err := api.Parse(r.Body, inf.Tx.Tx, &ds); err != nil {
+		api.HandleErr(w, r, inf.Tx.Tx, http.StatusBadRequest, errors.New("decoding: "+err.Error()), nil)
+		return tc.DeliveryServiceNullable{}, false
+	}
+
+	ok, err := updateDSSafe(inf.Tx.Tx, dsID, ds)
+	if err != nil {
+		api.HandleErr(w, r, inf.Tx.Tx, http.StatusInternalServerError, nil, errors.New("updating delivery service safe: "+err.Error()))
+		return tc.DeliveryServiceNullable{}, false
+	}
+	if !ok {
+		api.HandleErr(w, r, inf.Tx.Tx, http.StatusNotFound, nil, nil)
+		return tc.DeliveryServiceNullable{}, false
+	}
+
+	dses, userErr, sysErr, errCode := readGetDeliveryServices(inf.Params, inf.Tx, inf.User)
+	if userErr != nil || sysErr != nil {
+		api.HandleErr(w, r, inf.Tx.Tx, errCode, userErr, sysErr)
+		return tc.DeliveryServiceNullable{}, false
+	}
+	if len(dses) != 1 {
+		api.HandleErr(w, r, inf.Tx.Tx, http.StatusInternalServerError, nil, fmt.Errorf("delivery service safe update, read expected 1 delivery service, got %v", len(dses)))
+		return tc.DeliveryServiceNullable{}, false
+	}
+	return dses[0], true
+}
+
+// updateDSSafe updates the given delivery service in the database. Returns whether the DS existed, and any error.
+func updateDSSafe(tx *sql.Tx, dsID int, ds tc.DeliveryServiceSafeUpdate) (bool, error) {
+	q := `
+UPDATE deliveryservice SET
+display_name=$1,
+info_url=$2,
+long_desc=$3,
+long_desc_1=$4
+WHERE id = $5
+RETURNING id
+`
+	res, err := tx.Exec(q, ds.DisplayName, ds.InfoURL, ds.LongDesc, ds.LongDesc1, dsID)
```

Comment:
Based on this, since `Sanitize` is called by the CRUDer before the handler runs, a payload like `{"id":1, displayName:"1"}` will also set the DS's longDesc, longDesc1 and InfoURL to empty strings:

## Initial DS
### Request
```http
GET /api/1.3/deliveryservices/1 HTTP/1.1
User-Agent: python-requests/2.22.0
Accept-Encoding: gzip, deflate
Accept: */*
Connection: keep-alive
Cookie: mojolicious=...
```
### Response (non-"safe" fields omitted)
```
HTTP/1.1 200 OK
Content-Encoding: gzip
Content-Type: application/json
Set-Cookie: mojolicious=...; Path=/; Expires=Fri, 03 Jan 2020 23:49:36 GMT; Max-Age=3600; HttpOnly
X-Server-Name: traffic_ops_golang/
Date: Fri, 03 Jan 2020 22:49:36 GMT
Content-Length: 731

{ "response": [
	{
		"displayName": "Demo 1",
		"id": 1,
		"infoUrl": null,
		"longDesc": "Apachecon North America 2018",
		"longDesc1": null
	}
]}
```

## Modification
### Request
```http
PUT /api/1.3/deliveryservices/1/safe HTTP/1.1
User-Agent: python-requests/2.22.0
Accept-Encoding: gzip, deflate
Accept: */*
Connection: keep-alive
Cookie: mojolicious=...
Content-Length: 68

{"id": 1, "displayName": "testquest", "infoUrl": "http://google.com"}
```

### Response (non-"safe" fields omitted)
```http
HTTP/1.1 200 OK
Content-Encoding: gzip
Content-Type: application/json
Set-Cookie: mojolicious=...; Path=/; Expires=Fri, 03 Jan 2020 23:50:42 GMT; Max-Age=3600; HttpOnly
X-Server-Name: traffic_ops_golang/
Date: Fri, 03 Jan 2020 22:50:42 GMT
Content-Length: 766

{ "alerts": [
	{
		"text": "Deliveryservice safe update was successful.",
		"level": "success"
	}
], "response": [
	{
		"displayName": "testquest",
		"id": 1,
		"infoUrl": "http://google.com",
		"longDesc": "",
		"longDesc1": "",
	}
]}
```

Reply:
That appears to be how the Perl works:
```
	my $values = {
		display_name           => $params->{displayName},
		info_url               => $params->{infoUrl},
		long_desc              => $params->{longDesc},
		long_desc_1            => $params->{longDesc1},
	};

	my $rs = $ds->update($values);
```
If any of those fields were missing in the Perl request they'd likely be set to null, right (if allowed)? I think the only difference here is that they'd be set to empty strings instead, which seems fine to me.

Reply:
Well in that case the docs need to be updated, because that's not what they say it does.